server/leafnode.go

```diff
@@ -0,0 +1,1074 @@
+		opts.LeafNode.Port = l.Addr().(*net.TCPAddr).Port
+	}
+
+	s.leafNodeInfo = info
+	// Possibly override Host/Port and set IP based on Cluster.Advertise
+	if err := s.setLeafNodeInfoHostPortAndIP(); err != nil {
+		s.Fatalf("Error setting leafnode INFO with LeafNode.Advertise value of %s, err=%v", s.opts.LeafNode.Advertise, err)
+		l.Close()
+		s.mu.Unlock()
+		return
+	}
+
+	// Generate the INFO JSON.
+	// FIXME(dlc) - If we always generate no need for this.
+	b, _ := json.Marshal(s.leafNodeInfo)
+	pcs := [][]byte{[]byte("INFO"), b, []byte(CR_LF)}
+	s.leafNodeInfoJSON = bytes.Join(pcs, []byte(" "))
+
+	// Setup state that can enable shutdown
+	s.leafNodeListener = l
+	s.mu.Unlock()
+
+	// Let them know we are up
+	close(ch)
+	ch = nil
+
+	tmpDelay := ACCEPT_MIN_SLEEP
+
+	for s.isRunning() {
+		conn, err := l.Accept()
+		if err != nil {
+			tmpDelay = s.acceptError("LeafNode", err, tmpDelay)
+			continue
+		}
+		tmpDelay = ACCEPT_MIN_SLEEP
+		s.startGoRoutine(func() {
+			s.createLeafNode(conn, nil)
+			s.grWG.Done()
+		})
+	}
+	s.Debugf("Leafnode accept loop exiting..")
+	s.done <- true
+}
+
+// RegEx to match a creds file with user JWT and Seed.
+var credsRe = regexp.MustCompile(`\s*(?:(?:[-]{3,}[^\n]*[-]{3,}\n)(.+)(?:\n\s*[-]{3,}[^\n]*[-]{3,}\n))`)
+
+// Lock should be held entering here.
+func (c *client) sendLeafConnect(tlsRequired bool) {
+	// We support basic user/pass and operator based user JWT with signatures.
+	cinfo := leafConnectInfo{
+		TLS:  tlsRequired,
+		Name: c.srv.info.ID,
+	}
+
+	// Check for credentials first, that will take precedence..
+	if creds := c.leaf.remote.Credentials; creds != "" {
+		c.Debugf("Authenticating with credentials file %q", c.leaf.remote.Credentials)
+		contents, err := ioutil.ReadFile(creds)
+		if err != nil {
+			c.Errorf("%v", err)
+			return
+		}
+		defer wipeSlice(contents)
+		items := credsRe.FindAllSubmatch(contents, -1)
+		if len(items) < 2 {
+			c.Errorf("Credentials file malformed")
+			return
+		}
+		// First result should be the user JWT.
+		// We copy here so that the file containing the seed will be wiped appropriately.
+		raw := items[0][1]
+		tmp := make([]byte, len(raw))
+		copy(tmp, raw)
+		// Seed is second item.
+		kp, err := nkeys.FromSeed(items[1][1])
+		if err != nil {
+			c.Errorf("Credentials file has malformed seed")
+			return
+		}
+		// Wipe our key on exit.
+		defer kp.Wipe()
+
+		sigraw, _ := kp.Sign(c.nonce)
+		sig := base64.RawURLEncoding.EncodeToString(sigraw)
+		cinfo.JWT = string(tmp)
+		cinfo.Sig = sig
+	} else if userInfo := c.leaf.remote.URL.User; userInfo != nil {
+		cinfo.User = userInfo.Username()
+		pass, _ := userInfo.Password()
+		cinfo.Pass = pass
+	}
+
+	b, err := json.Marshal(cinfo)
+	if err != nil {
+		c.Errorf("Error marshaling CONNECT to route: %v\n", err)
+		c.closeConnection(ProtocolViolation)
+		return
+	}
+	c.sendProto([]byte(fmt.Sprintf(ConProto, b)), true)
+}
+
+// Called when an inbound leafnode connection is either accepted.
```

Comment:
Comment seem wrong. This is called when accepting inbound or soliciting outbound I would think

Reply:
will fix.